pkg/build/controller/buildpod/controller.go

```diff
@@ -0,0 +1,404 @@
+	storeSyncedPollPeriod = 100 * time.Millisecond
+	maxRetries            = 60
+)
+
+// BuildPodController watches pods running builds and manages the build state
+type BuildPodController struct {
+	buildUpdater buildclient.BuildUpdater
+	secretClient kcoreclient.SecretsGetter
+
+	queue workqueue.RateLimitingInterface
+
+	buildStore oscache.StoreToBuildLister
+	podStore   cache.StoreToPodLister
+
+	buildStoreSynced func() bool
+	podStoreSynced   func() bool
+
+	runPolicies []policy.RunPolicy
+}
+
+// NewBuildPodController creates a new BuildPodController.
+func NewBuildPodController(buildInformer, podInformer cache.SharedIndexInformer, kc kclientset.Interface, oc osclient.Interface) *BuildPodController {
+	buildListerUpdater := buildclient.NewOSClientBuildClient(oc)
+	c := &BuildPodController{
+		buildUpdater: buildListerUpdater,
+		secretClient: kc.Core(), // TODO: Replace with cache client
+		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	c.runPolicies = policy.GetAllRunPolicies(buildListerUpdater, buildListerUpdater)
+
+	c.buildStore.Indexer = buildInformer.GetIndexer()
+	c.podStore.Indexer = podInformer.GetIndexer()
+	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.updatePod,
+		DeleteFunc: c.deletePod,
+	})
+	buildInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addBuild,
+		UpdateFunc: c.updateBuild,
+	})
+
+	c.buildStoreSynced = buildInformer.HasSynced
+	c.podStoreSynced = podInformer.HasSynced
+
+	return c
+}
+
+// Run begins watching and syncing.
+func (c *BuildPodController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	// Wait for the build store to sync before starting any work in this controller.
+	if !cache.WaitForCacheSync(stopCh, c.buildStoreSynced, c.podStoreSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down build pod controller")
+}
+
+// HandlePod updates the state of the build based on the pod state
+func (bc *BuildPodController) HandlePod(pod *kapi.Pod) error {
+	glog.V(5).Infof("Handling update of build pod %s/%s", pod.Namespace, pod.Name)
+	build, exists, err := bc.getBuildForPod(pod)
+	if err != nil {
+		glog.V(4).Infof("Error getting build for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+		return err
+	}
+	if !exists || build == nil {
+		glog.V(5).Infof("No build found for pod %s/%s", pod.Namespace, pod.Name)
+		return nil
+	}
+
+	nextStatus := build.Status.Phase
+	currentReason := build.Status.Reason
+
+	if build.Status.Phase != buildapi.BuildPhaseFailed {
+		switch pod.Status.Phase {
+		case kapi.PodRunning:
+			// The pod's still running
+			build.Status.Reason = ""
+			build.Status.Message = ""
+			nextStatus = buildapi.BuildPhaseRunning
+
+		case kapi.PodPending:
```

Comment:
Sure, I just felt it's annoying not to handle these cases in their logical order. Anyway a nit